### PR TITLE
[WIP] Refs #29891 -- Add rudimentary ANSI SQL quoting to parameters in queryset.query.

### PR DESCRIPTION
**Pull request by FlipperPA**
See this ticket for background: https://code.djangoproject.com/ticket/29891

This is an initial attempt to add rudimentary ANSI SQL quoting to parameters in `queryset.query`. This won't be 100% accurate, but should be an improvement over unquoted dates and strings.

I am submitting this as a work in progress to see if I'm on the right path, and get some feedback and ideas.

I'm unsure why the initial query built by `Query()` includes the `ESCAPE` clause.

### CODE REVIEW

File: django/db/models/sql/query.py [hunk 1]
```diff
@@ -260,10 +261,27 @@ def __str__(self):
...
 
         Parameter values won't necessarily be quoted correctly, since that is
-        done by the database interface at execution time.
+        done by the database interface at execution time; rudimentary quoting
+        is performed below.
         """
         sql, params = self.sql_with_params()
-        return sql % params
+
+        unquoted_param_types = (int, float, Decimal)
```
Comment (FlipperPA) [inline django/db/models/sql/query.py]: Should we add `complex` to this tuple?

File: django/db/models/sql/query.py [hunk 2]
```diff
@@ -260,10 +261,27 @@ def __str__(self):
...
+            if type(param) in unquoted_param_types:
+                quoted_params.append(param)
+            else:
+                if type(param) is str:
+                    param = param.replace("'", "''")
+
+                quoted_params.append("'{param}'".format(
+                    param=param,
+                ))
+        sql = sql.replace(" ESCAPE '\\'", "")
```
Comment (FlipperPA) [inline django/db/models/sql/query.py]: What does this `ESCAPE` do? I believe removing it is causing several tests in the suite to fail. I have this here, for now, to produce SQL that can be run.

Comment (timgraham) [reply django/db/models/sql/query.py]: It looks like it's from:
https://github.com/django/django/blob/c1c68d1ac0f0d50eb37df32892b132f31a1179da/django/db/backends/sqlite3/base.py#L101-L103
I guess that reveals one problem with expecting you can get SQL to run using this method (if you have a mixed multi-database configuration). `sql_with_params()` is called with `DEFAULT_DB_ALIAS`.

I don't have much expertise here but are we basically "reimplementing the full quoting behavior of a DB cursor" (or a simpler version) as mentioned in https://code.djangoproject.com/ticket/12611#comment:1.

**Comment by timgraham** [comment]
By the way, I tested a more complicated example:
```
>>> print(RangesModel.objects.filter(ints__exact=NumericRange(0, 10)).query)
SELECT ... WHERE "postgres_tests_rangesmodel"."ints" = NumericRange(0, 10, '[)')
```
This isn't close to being correct. It looks like the parameter is just being cast to a string. I guess it might be psycopg2 that does the conversion of NumericRange to SQL.

**Comment by FlipperPA** [comment]
Just an update: I'm going to marshal some resources to help out with this with a bit more experience in the ORM than me. We'll start with some failing tests and move forward from there. Thanks for the example, Tim.

**Comment by FlipperPA** [comment]
At DjangoCon US sprints, we've refactored and started with a new concept. Big thanks to @louiseGrandjonc for the work! We're looking to get this passing the test suite and then adding a test to test PostgreSQL's response.

**Comment by FlipperPA** [comment]
It looks like we have (at least) five related test failures we'll have to figure out:

```
======================================================================
FAIL [0.002s]: test_overridden_get_lookup (custom_lookups.tests.CustomisedMethodsTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-bionic/database/postgres/label/bionic-pr/python/python3.6/tests/custom_lookups/tests.py", line 547, in test_overridden_get_lookup
    self.assertIn('monkeys()', str(q.query))
AssertionError: 'monkeys()' not found in 'SELECT "custom_lookups_custommodel"."id", "custom_lookups_custommodel"."field" FROM "custom_lookups_custommodel" WHERE \'monkeys\'()'

======================================================================
FAIL [0.002s]: test_overridden_get_lookup_chain (custom_lookups.tests.CustomisedMethodsTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-bionic/database/postgres/label/bionic-pr/python/python3.6/tests/custom_lookups/tests.py", line 555, in test_overridden_get_lookup_chain
    self.assertIn('elephants()', str(q.query))
AssertionError: 'elephants()' not found in 'SELECT "custom_lookups_custommodel"."id", "custom_lookups_custommodel"."field" FROM "custom_lookups_custommodel" WHERE \'elephants\'()'

======================================================================
FAIL [0.002s]: test_overridden_get_transform (custom_lookups.tests.CustomisedMethodsTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-bionic/database/postgres/label/bionic-pr/python/python3.6/tests/custom_lookups/tests.py", line 551, in test_overridden_get_transform
    self.assertIn('banana()', str(q.query))
AssertionError: 'banana()' not found in 'SELECT "custom_lookups_custommodel"."id", "custom_lookups_custommodel"."field" FROM "custom_lookups_custommodel" WHERE \'banana\'() = \'3\''

======================================================================
FAIL [0.002s]: test_overridden_get_transform_chain (custom_lookups.tests.CustomisedMethodsTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-bionic/database/postgres/label/bionic-pr/python/python3.6/tests/custom_lookups/tests.py", line 559, in test_overridden_get_transform_chain
    self.assertIn('pear()', str(q.query))
AssertionError: 'pear()' not found in 'SELECT "custom_lookups_custommodel"."id", "custom_lookups_custommodel"."field" FROM "custom_lookups_custommodel" WHERE \'pear\'() = \'3\''

======================================================================
FAIL [0.002s]: test_in_keeps_value_ordering (lookup.tests.LookupTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-bionic/database/postgres/label/bionic-pr/python/python3.6/tests/lookup/tests.py", line 565, in test_in_keeps_value_ordering
    self.assertIn(' IN (a1, a2, a3, a4, a5, a6, a7) ', str(query))
AssertionError: ' IN (a1, a2, a3, a4, a5, a6, a7) ' not found in 'SELECT "lookup_article"."id" FROM "lookup_article" WHERE "lookup_article"."slug" IN (\'a1\', \'a2\', \'a3\', \'a4\', \'a5\', \'a6\', \'a7\') ORDER BY "lookup_article"."pub_date" DESC, "lookup_article"."headline" ASC'
```

**Comment by FlipperPA** [comment]
Hello @felixxm - Carlton asked me to tag you for this PR to the ORM.

@louiseGrandjonc and I worked on this during DjangoCon US sprints, to make the output of `queryset.query` valid SQL, starting with PostgreSQL. Since psycopg2 provides a `mogrify()` method on the cursor that outputs the actual SQL to be run against the database, this was a good starting point. If we want to add `mogrify` methods to other database engine cursors, this should "just work" for them as well.

We've run into a few test failures we needed your input on, listed above. When testing the output of the `mogrify` function directly, we can see it is quoting the function name in the SQL.
Example test: "tests/custom_lookups/tests.py", test_overridden_get_lookup

Old value:
SELECT "custom_lookups_custommodel"."id", "custom_lookups_custommodel"."field" FROM "custom_lookups_custommodel" WHERE monkeys()

New value:
SELECT "custom_lookups_custommodel"."id", "custom_lookups_custommodel"."field" FROM "custom_lookups_custommodel" WHERE 'monkeys'()

The quoting of `monkeys` above is causing these tests for fail. It looks to me that this is because `monkeys()` isn't being tested as a parameter, but as a function. I'm thinking we should probably perform a different test in this case, since `monkeys()` et al are not actually params. I wanted to run it by you to see:

(1) is this a show stopper, or can we just have the tests check for either `monkeys()` or `'monkeys'()` to get this across the finish line?
(2) any thoughts on how we could potentially rewrite the tests more appropriately, so we aren't treating functions as parameters in the test?

How can we proceed? Thanks for all of your efforts - sorry we didn't see you at DjangoCon US!

**Comment by louiseGrandjonc** [comment]
Hi @FlipperPA,

I did a few tests. And I think that one of the failing test, we can update it.

```
======================================================================
FAIL [0.002s]: test_in_keeps_value_ordering (lookup.tests.LookupTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-bionic/database/postgres/label/bionic-pr/python/python3.6/tests/lookup/tests.py", line 565, in test_in_keeps_value_ordering
    self.assertIn(' IN (a1, a2, a3, a4, a5, a6, a7) ', str(query))
AssertionError: ' IN (a1, a2, a3, a4, a5, a6, a7) ' not found in 'SELECT "lookup_article"."id" FROM "lookup_article" WHERE "lookup_article"."slug" IN (\'a1\', \'a2\', \'a3\', \'a4\', \'a5\', \'a6\', \'a7\') ORDER BY "lookup_article"."pub_date" DESC, "lookup_article"."headline" ASC'
```

The query generated by this patch is:
`SELECT "lookup_article"."id" FROM "lookup_article" WHERE "lookup_article"."slug" IN ('a1', 'a2', 'a3', 'a4', 'a5', 'a6', 'a7') ORDER BY "lookup_article"."pub_date" DESC, "lookup_article"."headline" ASC`

And it's actually what postgres would expect. 

The following query would raise a syntax error:
`SELECT "lookup_article"."id" FROM "lookup_article" WHERE "lookup_article"."slug" IN (a1, a2, a3, a4, a5, a6, a7) ORDER BY "lookup_article"."pub_date" DESC, "lookup_article"."headline" ASC`


My question would be: should we have two tests, one with 
` @unittest.skipUnless(connection.vendor == 'postgresql', "PostgreSQL specific SQL returned")`
and one with 
` @unittest.skipUnless(connection.vendor != 'postgresql', "Db not using psycopg2 mogrify")`

or here is an other solution I've written in my repo: https://github.com/louiseGrandjonc/django/commit/dff9ac87eeca24364d2d6dc1b3e6d80cf6630991


Best,
Louise

**Comment by FlipperPA** [comment]
That's great, @louiseGrandjonc! Let's wait until we heard back from @felixxm and @carltongibson for any answer on how to handle those tests as well. I think it may just take a few days to hear back, since they've been flooding with PRs from DjangoCon US.

Awesome work on getting that test outputting the proper SQL!

File: django/db/models/sql/query.py [hunk 3]
```diff
@@ -260,6 +260,11 @@ def __str__(self):
         done by the database interface at execution time.
         """
         sql, params = self.sql_with_params()
+
+        cursor = self.get_compiler(DEFAULT_DB_ALIAS).connection.cursor()
+        if hasattr(cursor, 'mogrify'):
+            return cursor.mogrify(sql, params).decode('utf-8')
+
```
[the remark this answers is not in the document]
Comment (felixxm) [reply django/db/models/sql/query.py]: IMO we should add to operations a new method e.g. `query_sql()` with base implementation (`BaseDatabaseOperations`):
```python
def query_sql(self, sql, params):
    return sql % params
```
and use `mogrify()` on PostgreSQL:
```python
def query_sql(self, sql, params):
    with self.connection.cursor() as cursor:
        return cursor.mogrify(sql, params).decode()
```

**Comment by FlipperPA** [comment]
Thank you all for the review and comments. I'm at PyGotham but will review these once I'm back and caught up. I just saw a talk here which recommended looking at `queryset.query` output to see the SQL being built for Q objects, so apparently, this patch will help out more folks than just me and my obsessive/compulsive nature!

**Comment by felixxm** [comment]
I moved `SQLFuncMixin.as_sql()` fix to a separate PR #12168. We also have an alternative approach proposed by Claude #12156.

**Comment by felixxm** [comment]
@FlipperPA @louiseGrandjonc Many thanks for a patch and discussion :rocket: :1st_place_medal:   The main issue with getting an executable SQL from `QuerySet.query` is related with adapting and quoting parameters. I think we should prefer Claude's (@claudep) approach (see #12156) because:
 - it's handle by the `psycopg2` and it doesn't require a new database cursor unlike `mogrify()` (that's why had to change the `SimpleTestCase` to `TestCase` in few cases),
- it should be easier to implement similar solution for other back-ends.

**Pull request closed without merging**